### PR TITLE
[Tune] Raise better error messages for failures with `ResultGrid.get_best_result`

### PR DESCRIPTION
**Pull request by amogkam**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by Yard1 — APPROVED** [review]

File: python/ray/tune/tests/test_result_grid.py [hunk 1]
```diff
@@ -36,8 +37,41 @@ def f(config):
...
+    assert best_result.metrics["x"] == 2
+
+
+def test_best_result_no_report():
+    def f(config):
+        pass
+
+    analysis = tune.run(f, config={"x": tune.grid_search([1, 2])})
+    result_grid = ResultGrid(analysis)
+    with pytest.raises(RuntimeError):
```
Comment (xwjiang2010) [inline python/ray/tune/tests/test_result_grid.py:58]: nit, use `self.assertRaisesRegex`?

File: python/ray/tune/tests/test_result_grid.py [hunk 2]
```diff
@@ -36,8 +37,41 @@ def f(config):
...
+
+
+def test_no_metric_mode():
+    def f(config):
+        tune.report(x=1)
+
+    analysis = tune.run(f)
+    result_grid = ResultGrid(analysis)
+    with pytest.raises(ValueError):
+        result_grid.get_best_result()
```
Comment (xwjiang2010) [inline python/ray/tune/tests/test_result_grid.py:69]: Should we separate them? Since all should raise an error?

**Review by xwjiang2010 — APPROVED** [review]
Thank you!!

File: python/ray/tune/tests/test_result_grid.py (hunk 1 above)
Comment (amogkam) [reply python/ray/tune/tests/test_result_grid.py]: Updated!

File: python/ray/tune/tests/test_result_grid.py (hunk 2 above)
Comment (amogkam) [reply python/ray/tune/tests/test_result_grid.py:69]: Ah yeah good point...updated!

File: python/ray/tune/result_grid.py [hunk 3]
```diff
@@ -71,14 +71,39 @@ def get_best_result(
...
+        if not metric and not self._experiment_analysis.default_metric:
+            raise ValueError(
+                "No metric is provided. Either pass in a `metric` arg to "
+                "`get_best_result` or specify a metric in the "
+                "`TuneConfig` of your `Tuner`."
             )
+        if not mode and not self._experiment_analysis.default_mode:
+            raise ValueError(
+                "No mode is provided. Either pass in a `metric` arg to "
+                "`get_best_result` or specify a metric in the "
```
Comment (krfricke) [inline python/ray/tune/result_grid.py:83]:
```suggestion
                "No mode is provided. Either pass in a `mode` arg to "
                "`get_best_result` or specify a mode in the "
```

File: python/ray/tune/result_grid.py [hunk 4]
```diff
@@ -71,14 +71,39 @@ def get_best_result(
...
+        best_trial = self._experiment_analysis.get_best_trial(
+            metric=metric,
+            mode=mode,
+            scope=scope,
+            filter_nan_and_inf=filter_nan_and_inf,
         )
+        if not best_trial:
+            error_msg = (
+                "No best trial found for the given metric: "
+                f"{metric if metric else self._experiment_analysis.default_metric}. "
```
Comment (krfricke) [inline python/ray/tune/result_grid.py:95]:
```suggestion
                f"{metric or self._experiment_analysis.default_metric}. "
```

File: python/ray/tune/result_grid.py [hunk 5]
```diff
@@ -71,14 +71,39 @@ def get_best_result(
...
+                "This means that no trial has reported this metric"
+            )
+            error_msg += (
+                ", or all values reported for this metric is NaN. To not ignore NaN "
```
Comment (krfricke) [inline python/ray/tune/result_grid.py:99]:
```suggestion
                ", or all values reported for this metric are NaN. To not ignore NaN "
```

**Review by krfricke — APPROVED** [review]
LGTM just a few nits

**Pull request merged**